pkg/webhook/configvalidator/handler.go

```diff
@@ -0,0 +1,201 @@
+type Handler struct {
+	Logger    logr.Logger
+	APIReader client.Reader
+	Client    client.Reader
+	Decoder   admission.Decoder
+
+	ConfigMapKind               string
+	ConfigMapDataKey            string
+	GetConfigMapNameFromShoot   func(shoot *gardencore.Shoot) string
+	SkipValidationOnShootUpdate func(shoot, oldShoot *gardencore.Shoot) bool
+	AdmitConfig                 func(configRaw string, shootsReferencingConfigMap []*gardencore.Shoot) (int32, error)
+}
+
+// Handle validates configuration part of ConfigMaps which are referenced in Shoot resources.
+func (h *Handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	requestGK := schema.GroupKind{Group: req.Kind.Group, Kind: req.Kind.Kind}
+
+	switch requestGK {
+	case schema.GroupKind{Group: gardencorev1beta1.GroupName, Kind: "Shoot"}:
+		return h.admitShoot(ctx, req)
+	case schema.GroupKind{Group: corev1.GroupName, Kind: "ConfigMap"}:
+		return h.admitConfigMap(ctx, req)
+	}
+
+	return admissionwebhook.Allowed("resource is neither of type *core.gardener.cloud/v1beta1.Shoot nor *corev1.ConfigMap")
+}
+
+func (h *Handler) admitShoot(ctx context.Context, request admission.Request) admission.Response {
+	shoot := &gardencore.Shoot{}
+	if err := runtime.DecodeInto(gardenCoreDecoder, request.Object.Raw, shoot); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if shoot.DeletionTimestamp != nil {
+		// don't validate shoot if it's already marked for deletion, otherwise gardener-apiserver will deny the user's/
+		// controller's request, because we changed the spec
+		return admissionwebhook.Allowed("shoot is already marked for deletion")
+	}
+
+	configMapName := h.GetConfigMapNameFromShoot(shoot)
+	if configMapName == "" {
+		return admissionwebhook.Allowed(fmt.Sprintf("Shoot resource does not specify any %s ConfigMap", h.ConfigMapKind))
+	}
+
+	var oldShoot *gardencore.Shoot
+	if request.Operation == admissionv1.Update {
+		oldShoot = &gardencore.Shoot{}
+		if err := runtime.DecodeInto(gardenCoreDecoder, request.OldObject.Raw, oldShoot); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		// skip verification if spec wasn't changed
+		// this way we make sure, that users/gardenlet can always annotate/label the shoot if the spec doesn't change
+		if apiequality.Semantic.DeepEqual(oldShoot.Spec, shoot.Spec) {
+			return admissionwebhook.Allowed("shoot spec was not changed")
+		}
+	}
+
+	// `oldConfigMapName` is empty for CREATE shoot requests that specify a ConfigMap reference, hence we can skip the
+	// validation.
+	// Additionally, we only need to revalidate if the ConfigMap data is compatible with a new Kubernetes version if the
+	// version changed.
+	if oldShoot != nil &&
+		h.GetConfigMapNameFromShoot(oldShoot) == configMapName &&
+		oldShoot.Spec.Kubernetes.Version == shoot.Spec.Kubernetes.Version &&
+		(h.SkipValidationOnShootUpdate == nil || h.SkipValidationOnShootUpdate(shoot, oldShoot)) {
```

Comment:
On a second read, I think we skip some validation cases here that we wouldn't want to skip. For instance, the `authorizationconfig` handler does not set a `h.SkipValidationOnShootUpdate` function, so if the ConfigMap name didn't change and the k8s version didn't change, but the `shoot.spec.kubernetes.kubeAPIServer.StructuredAuthorization.Kubeconfigs` changed, the validation would still be skipped.


I guess, `h.SkipValidationOnShootUpdate` should be set for the `authorizationconfig` handler, similar to the `authenticationconfig` one.

---

Alternatively, as a personal preference, I think it makes a bit more sense to not skip the validation if `h.SkipValidationOnShootUpdate` is not set, so what about flipping this check (bad thing is that we will then have to add a dummy function that always returns true for the audit policy handler):
```suggestion
		(h.SkipValidationOnShootUpdate != nil && h.SkipValidationOnShootUpdate(shoot, oldShoot)) {
```

Reply:
> On a second read, I think we skip some validation cases here that we wouldn't want to skip. For instance, the `authorizationconfig` handler does not set a `h.SkipValidationOnShootUpdate` function, so if the ConfigMap name didn't change and the k8s version didn't change, but the `shoot.spec.kubernetes.kubeAPIServer.StructuredAuthorization.Kubeconfigs` changed, the validation would still be skipped.
> 
> I guess, `h.SkipValidationOnShootUpdate` should be set for the `authorizationconfig` handler, similar to the `authenticationconfig` one.

Good suggestion, thank you! :) Done!

> Alternatively, as a personal preference, I think it makes a bit more sense to not skip the validation if `h.SkipValidationOnShootUpdate` is not set, so what about flipping this check (bad thing is that we will then have to add a dummy function that always returns true for the audit policy handler):

I don't want to specify this function when not needed - that's why I prefer the current approach. We still only skip both `ConfigMap` name and Kubernetes version have not changed. Either the function is not set, or it is set and tells us to still skip. I don't understand why you prefer it the other way round with a dummy function 😕

Reply:
> I don't understand why you prefer it the other way round with a dummy function

I was making the analogy that the default value for a boolean is false -> this is a function that returns a boolean, so if the function itself is not set, the default behaviour should also be "false". But from a practical PoV you're right that having to add a dummy function is not too good, so I'm ok with keeping it as it is :)

Reply:
OK, great. Thank you :)